### ISSUE TITLE
Race between reading refs with libgit2 and packing refs with git (plus `rmdir refs/heads`)

### ISSUE THREAD

**Issue by spraints** [libgit2/libgit2#3412]
When one process packs refs (and removes the `refs/heads` dir) and another uses libgit2 to list refs, the libgit2 process experiences an error. I was able to reproduce it with a [couple of bash and ruby scripts](https://gist.github.com/spraints/07363dde0a6a7ba122ba).

**Comment by carlosmn** [comment]
What version are you on? We made the iterators ignore disappearing directories a while back. It was backported in `a4ded3` which was included in the 0.23.1 maintenance release.

**Comment by vmg** [comment]
@carlosmn I believe @spraints picked up that fix -- this race seems to be different than the object repacking one we spotted. Which is weird, because afaict the refs code also uses those iterators...

**Comment by carlosmn** [comment]
That's why I'm asking, yeah. That commit should have made it work for anything that wants to iterate the filesystem.

**Comment by spraints** [comment]
D'oh sorry for the noise. I wrote the gist against https://github.com/libgit2/rugged/commit/edcaca9. I tried it again with 0.23.2, and got an error, though it was a little bit different ("Could not read directory '': No such file or directory" instead of "Failed to open directory '.../07363dde0a6a7ba122ba/test.git//refs/heads/': No such file or directory"). I just checked the gist on Linux with rugged 0.23.2, and it looks fixed.
